### PR TITLE
 Refactors explosions

### PR DESCRIPTION
**Pull request by Cyberboss**
# CAN'T ESCAPE THE ADMIN BOMB!!!

:cl:
experiment: Explosions will no longer have a start up delay
/:cl:

### CODE REVIEW

**Comment by Iamgoofball** [comment]
wait a minute we can do multithreading???

we need to apply this everywhere

File: code/datums/explosion.dm [hunk 1]
```diff
@@ -1,10 +1,48 @@
...
+	if(TICK_CHECK) {\
+		stoplag();\
+		EX_PREPROCESS_EXIT_CHECK\
+	}
+
+/datum/explosion/New(atom/epicenter, devastation_range, heavy_impact_range, light_impact_range, flash_range, adminlog, ignorecap, flame_range, silent, smoke)
+	set waitfor = FALSE
+
+	var/id = ++id_counter
+	src.id = id
```
Comment (Iamgoofball) [inline code/datums/explosion.dm]: >src

**Comment by Cyberboss** [comment]
>BYOND multithreading

File: code/datums/explosion.dm (hunk 1 above)
Comment (Cyberboss) [inline code/datums/explosion.dm]: intentional to keep id as a proc var

File: code/datums/explosion.dm [hunk 2]
```diff
@@ -48,90 +85,79 @@
...
+
+	if(reactionary)
+		cached_exp_block = CaculateExplosionBlock(affected_turfs)
+
+	//lists are guaranteed to contain at least 1 turf at this point
+
+	var/Iteration = 0
+	var/AffTurfLen = affected_turfs.len
+	var/ExpBlockLen = cached_exp_block.len
+	for(var/turf/T in affected_turfs)
```
Comment (Iamgoofball) [inline code/datums/explosion.dm]: typeless lists

Comment (Cyberboss) [inline code/datums/explosion.dm]: Wow, that was a leftover, can't believe it went unnoticed till now

File: code/datums/explosion.dm [hunk 3]
```diff
@@ -177,24 +237,68 @@
...
+			var/current_exp_block = T.density ? T.explosion_block : 0
 
-/proc/secondaryexplosion(turf/epicenter, range)
-	for(var/turf/tile in spiral_range_turfs(range, epicenter))
-		tile.ex_act(2)
+			for(var/obj/machinery/door/D in T)
+				if(D.density)
+					current_exp_block += D.explosion_block
 
+			for(var/obj/structure/window/W in T)
```
Comment (Iamgoofball) [inline code/datums/explosion.dm]: typeless loops

**Comment by Iamgoofball** [comment]
>talking about doing multithreading

how do i do this you memelord

**Comment by Cyberboss** [comment]
It's just 3 procs with `set waitfor = FALSE` that understand how sleeps, references, and return values work and abuse them and work together on different parts of the same task (turf gathering, density calculation, and actual explosion).

File: code/datums/explosion.dm (hunk 3 above)
Comment (Cyberboss) [inline code/datums/explosion.dm]: Good use, it's looking specifically for windows in the turf

**Comment by Iamgoofball** [comment]
Ok

File: code/datums/explosion.dm [hunk 4]
```diff
@@ -177,24 +238,68 @@
...
+			A.sense_explosion(epicenter, devastation_range, heavy_impact_range, light_impact_range, took,orig_dev_range, orig_heavy_range, orig_light_range)
+
+	++stopped
+	qdel(src)
 
-	return 1
+#undef EX_PREPROCESS_EXIT_CHECK
+#undef EX_PREPROCESS_CHECK_TICK
 
+//asyncly populate the affected_turfs list
```
Comment (xxalpha) [inline code/datums/explosion.dm]: Why

Comment (Cyberboss) [inline code/datums/explosion.dm]: Title

**Comment by Robustin** [comment]
Finally, its been an embarassment.

**Comment by Fox-McCloud** [comment]
Theory is nice and all, but it can lead to situations where actual impact is...bad. Take pooling, for example.

Would need to see actual benchmarks of this, before I'm convinced.

**Comment by Cyberboss** [comment]
It generally takes about 8 seconds at most to finish all the precalculations. After that, it has the performance of a regular explosion. Difference being, time spent before with no explosion happening now will have an explosion happening.

**Comment by Robustin** [comment]
I'd really like to see some round with this testmerged.

I'm extremely curious as to what it will do to Lavaland 22/44/88's

**Comment by Fox-McCloud** [comment]
Could you expand on what you mean? It's a bit unclear.

**Comment by Iamgoofball** [comment]
I think it's basically:

- Explosions dont lock the server up while processing how much shit goes kaboom

**Comment by Iamgoofball** [comment]
@Cyberboss How would I implement this into other procs, btw? Can you make a """multithreading""" for dummies?

**Comment by Cyberboss** [comment]
@Fox-McCloud Before all this stuff was done serially

- Run `spiral_range_turfs` (slow/blocking) for the range of the explosion and get the list of affected turfs
- Iterate over that list to calculate the explosion block value of that tile (tick checked)
By this point, a noticeable amount of may have passed for big explosions since the start
- Begin exploding the turfs

Now what happens is this:

while(not_all_turfs_exploded)
- Gather turfs in `spiral_range_turfs` until tick timeout
- Calculate explosion block for gathered turfs until tick timeout
- Explode prepared turfs until tick timeout

This means that, while the turfs are being prepared, they are also blowing up at the same time

**Comment by Fox-McCloud** [comment]
Right.

You know, I half wonder if this should be a subsystem instead of just a datum. Everything is moving in that direction, and it would allow for better integration, at the end of the day.

**Comment by optimumtact** [comment]
Maybe

File: code/datums/explosion.dm [hunk 5]
```diff
@@ -1,10 +1,48 @@
-/proc/explosion(turf/epicenter, devastation_range, heavy_impact_range, light_impact_range, flash_range, adminlog = 1, ignorecap = 0, flame_range = 0 ,silent = 0, smoke = 1)
-	set waitfor = 0
-	src = null //so we don't abort once src is deleted
+GLOBAL_LIST_EMPTY(explosions)
+//Against my better judgement, I will return the explosion datum
+//If I see any GC errors for it I will find you
+//and I will gib you
+//	--Cyberboss
```
Comment (optimumtact) [inline code/datums/explosion.dm]: no

File: code/datums/explosion.dm [hunk 6]
```diff
@@ -1,10 +1,48 @@
...
+/proc/explosion(atom/epicenter, devastation_range, heavy_impact_range, light_impact_range, flash_range, adminlog = TRUE, ignorecap = FALSE, flame_range = 0 , silent = FALSE, smoke = FALSE)
+	return new /datum/explosion(epicenter, devastation_range, heavy_impact_range, light_impact_range, flash_range, adminlog, ignorecap, flame_range, silent, smoke)
+
+//This datum creates 3 threads
```
Comment (optimumtact) [inline code/datums/explosion.dm]: they're more like tasks on a event loop

Comment (optimumtact) [inline code/datums/explosion.dm]: threads is confusing terminology to use here

File: code/datums/explosion.dm [hunk 7]
```diff
@@ -48,90 +85,80 @@
...
+		cached_exp_block = CaculateExplosionBlock(affected_turfs)
+
+	//lists are guaranteed to contain at least 1 turf at this point
+
+	var/Iteration = 0
+	var/AffTurfLen = affected_turfs.len
+	var/ExpBlockLen = cached_exp_block.len
+	for(var/TI in affected_turfs)
+		var/turf/T = TI
+		++Iteration
```
Comment (optimumtact) [inline code/datums/explosion.dm]: >capitalised var

File: code/datums/explosion.dm [hunk 8]
```diff
@@ -143,27 +170,61 @@
...
 
 		var/throw_dir = get_dir(epicenter,T)
 		for(var/obj/item/I in T)
-			if(I && !I.anchored)
+			if(!I.anchored)
 				var/throw_range = rand(throw_dist, max_range)
 				var/turf/throw_at = get_ranged_target_turf(I, throw_dir, throw_range)
 				I.throw_speed = 4 //Temporarily change their throw_speed for embedding purposes (Reset when it finishes throwing, regardless of hitting anything)
-				I.throw_at(throw_at, throw_range, I.throw_speed)
+				I.throw_at(throw_at, throw_range, 4)
```
Comment (optimumtact) [inline code/datums/explosion.dm]: DEFINE

File: code/datums/explosion.dm [hunk 9]
```diff
@@ -143,27 +170,61 @@
...
+			
+			//update the trackers
+			AffTurfLen = affected_turfs.len
+			ExpBlockLen = cached_exp_block.len
+
+			if(break_condition)
+				if(reactionary)
+					//until there are more block checked turfs than what we are currently at
+					//or the explosion has stopped
+					UNTIL(Iteration < AffTurfLen || !running)
```
Comment (optimumtact) [inline code/datums/explosion.dm]: More capitalised vars

File: code/datums/explosion.dm (hunk 3 above)
Comment (optimumtact) [inline code/datums/explosion.dm]: @Iamgoofball typeless loops only apply when the list is full of only one type you dip

File: code/datums/explosion.dm (hunk 5 above)
Comment (Cyberboss) [inline code/datums/explosion.dm]: No to my comment? or returning the datum?

Comment (optimumtact) [inline code/datums/explosion.dm]: your name

Comment (RemieRichards) [inline code/datums/explosion.dm]: good film

**Comment by Cyberboss** [comment]
@Fox-McCloud You were right

**Comment by Cyberboss** [comment]
Literally unplayable, nevermind

```
4:42:47 PM Cyberboss mso what has more tick time in general, subsystems or everything else? 
4:43:30 PM +MrStonedOne depends on what order byond resumes in 
4:43:58 PM +MrStonedOne generally the mc is the last thing to resume, and would only be allowed to run if the other things saved it some tick time 
4:44:36 PM +MrStonedOne but the mc increases how long it sleeps (since byond resumes in order something slept, things that have been sleeping for longer get resumed sooner) if it starts up late and doesn't have much of a tick to work with 
4:44:57 PM Cyberboss yeah, someone suggested I port explosions to an SS 
4:45:00 PM Cyberboss terrible idea 
4:45:04 PM Cyberboss lags it to hell 
4:45:18 PM * Joan claps 
4:45:40 PM +MrStonedOne ya, ideally it would equalize out to 50/50, but generally, for bursts of activity, non-ss will get a larger amount of time until the mc adjusts 
```

**Comment by optimumtact** [comment]
"""Denied"""

**Comment by Cyberboss** [comment]
Stale WHY FRUIT!

**Comment by optimumtact** [comment]
we should get a test merge

**Comment by Cyberboss** [comment]
tested

**Comment by Siserith** [comment]
bombs dont seem to be able to kill anyone... i made a bunch of bombs today and twice a group of three people were fine when a bomb detonated on the tile right next to them, then when science was bombed, a person right on top of it as it went of, was fine, and walked away from the crater mostly unharmed except by debris. and four people nearby, were fine except for debris, then later as i was running away and two people came and started beating me with bats, i sent a signal, and i quickly went into crit, the bombs went off and one of the people standing on top of me, and the other next to me when it went off, were fine, it blew up the first few tiles around me, slowed a bit as it got further out, and they just ran away. not even taking any damage from the blast. while i myself got instantly gibbed, penguin told me to put this here

**Comment by Cyberboss** [comment]
Ugh, something is sleeping in ex_act which is causing this!

**Comment by Cyberboss** [comment]
Problem was caused by the waitfor, I'll remove that bit

**Comment by optimumtact** [comment]
Will this supersede #27225

**Comment by Cyberboss** [comment]
Nope that's was actually the cause of the bug with this PR

**Comment by Dorsisdwarf** [comment]
Would this remove the slow-explosions issue where you can casually walk away faster than a bomb can go off?

**Comment by Cyberboss** [comment]
Nothing can really fix that since it's just lag from the sheer amount of
proc calls explosions incur

**Comment by Dorsisdwarf** [comment]
Yeah but I actually preferred the old "Entire station freezes up for a second and everyone panics to see if the bomb is going to go off nexty to them or far away" to "spacemen exist in superspeed hyperspace and can easily escape any big explosion"

**Comment by Iamgoofball** [comment]
Too bad then

**Comment by Fox-McCloud** [comment]
I'm still of the mindset that this should be an entire subsystem, with merely just iterating over the explosion datum, the things they'd imapct, then actually calling ex_act() ....every tick (with the tickrate being as fast as the lighting controller...which is as fast as the game ticks) and halting if the current elapsed tick is going to get used up.

**Comment by Cyberboss** [comment]
I tried that, apparently it was too much for the MC to handle because they
slowed down to shit. See
https://github.com/tgstation/tgstation/pull/26361/commits/952043e4b863d4a6832a7b74d0f96e8e0ee2005e

**Comment by duncathan** [comment]
@MrStonedOne this seems like your area of expertise, thoughts?

**Comment by MrStonedOne** [comment]
@Fox-McCloud this is almost guaranteed to run worst in terms of cpu time and in terms of how long it takes a large explosion to process.

But it does refactor away the lag that happens when a large explosion first happens, where it booms but nothing processes for a few seconds and then the explosion starts propagating. It sadly does that by starting the explosion while it's still looking for the turfs to explode. splitting the process up like that will have a bit of an overhead.

I doubt it will be much and over all it's worth it.

**Comment by Cyberboss** [comment]
Since everyone seems too afraid to merge this

Current explosions: https://youtu.be/Q345NaE2m3k?t=9

Refactored explosions: https://youtu.be/Q345NaE2m3k?t=196

**Comment by optimumtact** [comment]
We're not afraid, we just know it's by you

**Comment by lzimann** [comment]
what could go wrong

**Comment by optimumtact** [comment]
Literally everything

**Pull request merged**